GoogleDataTransport/GDTCORLibrary/GDTCORDatabase.m

```diff
@@ -0,0 +1,369 @@
+ *   version.
+ * @param stmtCache The statement cache.
+ */
+static void RunMigrations(sqlite3 *db,
+                          NSDictionary<NSNumber *, NSString *> *migrations,
+                          CFMutableDictionaryRef stmtCache) {
+  if (!db) {
+    GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"The database is closed");
+    return;
+  }
+  int userVersion = GetUserVersion(db);
+  __block int newUserVersion = userVersion;
+  [migrations enumerateKeysAndObjectsUsingBlock:^(NSNumber *_Nonnull key, NSString *_Nonnull obj,
+                                                  BOOL *_Nonnull stop) {
+    if (key.intValue > newUserVersion) {
+      if (RunNonQuery(db, nil, obj, stmtCache, NO)) {
+        newUserVersion = key.intValue;
+      } else {
+        GDTCORLogError(GDTCORMCEDatabaseError, @"%@", @"Migration failed: version:%@ statement:%@",
+                       key, obj);
+      }
+    }
+  }];
+
+  if (newUserVersion != userVersion) {
+    SetUserVersion(db, newUserVersion);
+  }
+}
+
+/** Creates and/or returns a static queue shared across the GDTCORDatabase class.
+ *
+ * @return The class-shared queue.
+ */
+static dispatch_queue_t SharedQueue() {
+  static dispatch_queue_t sharedQueue;
+  static dispatch_once_t onceToken;
+  dispatch_once(&onceToken, ^{
+    sharedQueue = dispatch_queue_create("com.google.GDTCORDatabase.shared", DISPATCH_QUEUE_SERIAL);
+  });
+  return sharedQueue;
+}
+
+/** A dispatch_once token to manage the instantiation of the dbPathToInstance map. */
+static dispatch_once_t cacheToken;
+
+/** A shared cache of sqlite paths db instances. */
+static NSMutableDictionary<NSString *, GDTCORDatabase *> *dbPathToInstance;
+
+/** Maps db paths to instances, or deletes the mapping.
+ *
+ * @param instance The instance to map.
+ * @param path The path to map the instance to.
+ * @param delete If YES, deletes the mapping.
+ */
+static void SetInstanceToFileMap(GDTCORDatabase *instance, NSString *path, BOOL delete) {
+  dispatch_once(&cacheToken, ^{
+    dbPathToInstance = [[NSMutableDictionary alloc] init];
+  });
+  // Don't map anything if it's a special sqlite path.
+  if ([path hasPrefix:@":"] && [path hasSuffix:@":"]) {
+    return;
+  }
+  dispatch_async(SharedQueue(), ^{
+    if (delete) {
+      [dbPathToInstance removeObjectForKey:path];
+    } else {
+      dbPathToInstance[path] = instance;
+    }
+  });
+}
+
+/** Returns YES if there's an instance in existence for the given sqlite path.
+ *
+ * @param path The path to check.
+ * @return YES, if there's a db in memory operating on that path already.
+ */
+static BOOL InstanceExistsForPath(NSString *path) {
+  // Only 1 instance per file is allowed. Return nil if one is already open (including :memory:).
+  dispatch_once(&cacheToken, ^{
+    dbPathToInstance = [[NSMutableDictionary alloc] init];
+  });
+
+  // Check if this is a special type of path. If it begins and ends with :, it's ok to create.
```

Comment:
I wonder what is a special type of path?

Reply:
From https://www.sqlite.org/c3ref/open.html

_If the filename is ":memory:", then a private, temporary in-memory database is created for the connection. This in-memory database will vanish when the database connection is closed. **Future versions of SQLite might make use of additional special filenames that begin with the ":" character**. It is recommended that when a database filename actually does begin with a ":" character you should prefix the filename with a pathname such as "./" to avoid ambiguity._

We use it mostly for testing, but probably tvOS eventually.

Reply:
Nice, thanks!